File: utils/Logger.js
/**
 * Logger
 * ------
 * The "logger" is really a log-broker.
 *
 * You register log-handler classes with it, which each implement
 * a log(message) method. When you log to this class, it forwards
 * the message to the registered handlers.
 *
 * This class manages log levels, formatting and indenting as well
 * as the collection of log-handlers.
 */

/**
 * @constructor
 */
function Logger() {
}

/**
 * An enum for log levels.
 */
Logger.LogLevel = {
    DEBUG: [0, "Debug"],
    INFO: [1, "Info"],
    INFO_PLUS: [3, "InfoPlus"],
    WARNING: [4, "Warning"],
    ERROR: [5, "Error"],
    FATAL: [6, "Fatal"]
};

// Indent level (sections of the log can be indented)...
Logger._indent_level = 0;

// The collection of log-handlers...
var set = require('collections/set');
Logger._handlers = new set();

/**
 * Adds a log-handler. This is an object implementing:
 * handle_log_message(message, level, indent_level)
 */
Logger.addHandler = function(handler) {
    Logger._handlers.add(handler);
};

/**
 * Removes a log handler.
 */
Logger.removeHandler = function(handler) {
    Logger._handlers.delete(handler);
};

/**
 * Logs the message.
 */
Logger.log = function(message, level) {
    // We find the indent...
    var indent = '';
    if(Logger._indent_level > 0) {
        indent = new Array(Logger._indent_level).join('  ') + '- ';
    }

    // We add the log-level if it is high enough...
    var prefix  = '';
    if(level[0] >= Logger.LogLevel.WARNING[0]) {
        prefix = level[1] + ': ';
    }

    // We log the formatted message...
    var formattedMessage = indent + prefix + message;
    Logger._handlers.forEach(function(handler) {
        handler.log(formattedMessage, level);
    });
};

/**
 * Increases the indent level.
 */
Logger.indent = function() {
    Logger._indent_level += 1;
};

/**
 * Decreases the indent level.
 */
Logger.dedent = function() {
    if(Logger._indent_level > 0) {
        Logger._indent_level -= 1;
    }
};

// Exports...
module.exports = Logger;

File: AIs/AI_BootAndShoot/BootAndShoot/Logger.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace BootAndShoot
{
    /// <summary>
    /// Logs messages to a log file.
    /// </summary>
    static class Logger
    {
        // An enum for log levels...
        public enum LogLevel
        {
            DEBUG,
            INFO,
            WARNING,
            ERROR,
            FATAL
        }

        /// <summary>
        /// Constructor
        /// </summary>
        static Logger()
        {
            Directory.CreateDirectory("log");
        }

        /// <summary>
        /// Sets the log level
        /// </summary>
        public static void setLogLevel(LogLevel logLevel)
        {
            m_logLevel = logLevel;
        }

        /// <summary>
        /// Logs the message.
        /// </summary>
        public static void log(string message, LogLevel logLevel)
        {
            if(logLevel < m_logLevel)
            {
                return;
            }

            var messageToLog = String.Format("{0},{1}: {2}", DateTime.Now.ToString(), logLevel.ToString(), message);
            var lines = new List<string>();
            lines.Add(messageToLog);

            //File.WriteAllLines("log/BootAndShoot.log", lines);
            File.AppendAllLines("log/BootAndShoot.log", lines);
        }

        // The current log level...
        private static LogLevel m_logLevel = LogLevel.INFO;
    }
}
